integration/fuse/src/main/java/alluxio/fuse/file/MemoryCacheFileInStream.java

```diff
@@ -0,0 +1,236 @@
+  }
+
+  @Override
+  public long remaining() {
+    return mStatus.getLength() - mPos;
+  }
+
+  @Override
+  public int positionedRead(long position, byte[] buffer, int offset, int length)
+      throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public long getPos() throws IOException {
+    return mPos;
+  }
+
+  @Override
+  public void seek(long pos) throws IOException {
+    mPos = pos;
+  }
+
+  @Override
+  public int read() throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public int read(byte[] b) throws IOException {
+    return read(b, 0, b.length);
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    int length = Math.min(len, b.length - off);
+    int read = 0;
+    while (length > 0) {
+      byte[] bytes = getCacheBytes(length);
+      if (bytes.length == 0) {
+        break;
+      }
+      int copyLen = Math.min(length, bytes.length);
+      System.arraycopy(bytes, 0, b, off + read, copyLen);
+      mPos += copyLen;
+      read += copyLen;
+      length -= copyLen;
+    }
+    return read == 0 ? -1 : read;
+  }
+
+  @Override
+  public int read(ByteBuffer buf) throws IOException {
+    return super.read(buf);
+  }
+
+  @Override
+  public int read(ByteBuffer byteBuffer, int off, int len) throws IOException {
+    byteBuffer.position(off);
+    int length = Math.min(len, byteBuffer.capacity() - off);
+    byte[] data = new byte[length];
+    int read = read(data);
+    if (read > 0) {
+      byteBuffer.put(data, 0, read);
+    }
+    return read;
+  }
+
+  @Override
+  public void unbuffer() {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public long skip(long n) throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public int available() throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void close() throws IOException {
+    mFileInStream.close();
+  }
+
+  @Override
+  public synchronized void mark(int readlimit) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public synchronized void reset() throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public boolean markSupported() {
+    throw new UnsupportedOperationException();
+  }
+
+  private byte[] getCacheBytes(int length) throws IOException {
+    long posKey = (mPos / CACHE_SIZE) * CACHE_SIZE;
+    int start = (int) (mPos % CACHE_SIZE);
+    try {
+      if (mStatus.getLength() <= mPos) {
+        return new byte[0];
+      }
+      byte[] bytes = CACHE.get(
+          new CacheKey(mStatus.getPath(), mStatus.getLastModificationTimeMs(), posKey),
+          () -> {
+            mFileInStream.seek(posKey);
+            byte[] data = new byte[(int) Math.min(CACHE_SIZE, mStatus.getLength() - posKey)];
+            IOUtils.readFully(mFileInStream, data);
+            return data;
```

Comment:
wonder the details of race reads for a same 4M cache chunk, if 2 reads are happening to populate the cache, will this readFully happen at same time to worker and one of the data is discarded? or one thread is waiting for another thread to load the data and get its result?

Reply:
When two threads need to load the same data simultaneously, only one of the threads will execute the actual loading logic, while the other thread will wait.